client/block.c

```diff
@@ -1540,6 +1543,61 @@ const char* xdag_get_block_state_info(uint8_t flags)
 	return "Pending";
 }
 
+void append_block_info(struct block_internal *bi)
+{
+	// if websocket service is not running return directly
+	if(!g_websocket_running) {
+		return;
+	}
+
+	char time_buf[64] = {0};
+	char address[33] = {0};
+	uint64_t *h = bi->hash;
+	xdag_hash2address(h, address);
+	xdag_xtime_to_string(bi->time, time_buf);
+
+	char message[4096] = {0};
+	char buf[128] = {0};
+
+	sprintf(message,
+			"{\"time\":\"%s\""
+			",\"flags\":\"%x\""
+			",\"state\":\"%s\""
+			",\"hash\":\"%016llx%016llx%016llx%016llx\""
+			",\"difficulty\":\"%llx%016llx\""
+			",\"remark\":\"%s\""
+			",\"address\":\"%s\""
+			",\"balance\":\"%u.%09u\""
+			",\"fields\":["
+			, time_buf
+			, bi->flags & ~BI_OURS
+			, xdag_get_block_state_info(bi->flags)
+			, (unsigned long long)h[3], (unsigned long long)h[2], (unsigned long long)h[1], (unsigned long long)h[0]
+			, xdag_diff_args(bi->difficulty)
+			, get_remark(bi)
+			, address
+			, pramount(bi->amount)
+			);
+
+	if((bi->flags & BI_REF) && bi->ref != NULL) {
```

Comment:
there can be a segfault here

Reply:
Do you mean the old issue of the extra blocks? I think this function only called in block.c in the same thread as add_block_nolock.

Reply:
there are many threads in block.c that can change those values, and this function isn't executed by some thread about websocket?

Reply:
Not in some thread about web socket, but still in the same thread of add_block_nolock. web socket thread will fetch message from a queue which is protected by mutex.

Reply:
@sgaragagghu  Updated.